STM32F405RGTx/Core/Src/rtos_threads/state_machine.cpp

```diff
@@ -0,0 +1,440 @@
+    for (int i = 0; i < NUM_MOSFETS; ++i)
+    {
+        float mosfetTemp = global_mc_data.fet_temps[i];
+        if (mosfetTemp > MAX_MOSFET_TEMP_SEVERE)
+        {
+            ++temp_faults;
+        }
+        if (temp_faults > MIN_TEMP_FAULTS)
+        {
+            // TODO: add error code
+            return SevereDangerFault;
+        }
+    }
+    // check undervolts and overvolts for phase outputs
+    if (global_mc_data.pVa < MIN_VOLTAGE_SEVERE) {
+        undervolt_faults++;
+    }
+    if (global_mc_data.pVb < MIN_VOLTAGE_SEVERE) {
+        undervolt_faults++;
+    }
+    if (global_mc_data.pVc < MIN_VOLTAGE_SEVERE) {
+        undervolt_faults++;
+    }
+
+    if (global_mc_data.pVa > MAX_VOLTAGE_SEVERE) {
+        overvolt_faults++;
+    }
+    if (global_mc_data.pVb > MAX_VOLTAGE_SEVERE) {
+        overvolt_faults++;
+    }
+    if (global_mc_data.pVc > MAX_VOLTAGE_SEVERE) {
+        overvolt_faults++;
+    }
+
+    //  TODO: after checking for faults return state
+    if (overvolt_faults > MIN_OVERVOLT_FAULTS) {
+        return NormalDangerFault;
+    }
+    if (undervolt_faults > MIN_UNDERVOLT_FAULTS) {
+        return NormalDangerFault;
+    }
+
+    // Check current for phase outputs
+    if (global_mc_data.pIa > MAX_CURRENT_SEVERE || global_mc_data.pIb > MAX_CURRENT_SEVERE || global_mc_data.pIc > MAX_CURRENT_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+
+    // DC fault checking
+    if (global_mc_data.dc_voltage > MAX_DCVOLTAGE_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+    else if (global_mc_data.dc_voltage < MIN_DCVOLTAGE_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+    if (global_mc_data.dc_cap_temp > MAX_DCCAP_TEMP_SEVERE) {
+        return SevereDangerFault;
+    }
+    return NoFault;
+}
+
+State_t StateMachineThread::InitializeEvent(void) {
+    return Idle;
+}
+
+State_t StateMachineThread::IdleEvent(void) {
+    SetLedColour(0.0, 50.0, 0.0);
+    State_t normal_fault_check = NormalFaultChecking();
+    State_t severe_fault_check = SevereFaultChecking();
+    if (severe_fault_check == SevereDangerFault) {
+        return severe_fault_check;
+    }
+    else if (normal_fault_check == NormalDangerFault) {
+        return normal_fault_check;
+    }
+
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANZThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        idle_state_id = state_id;
+        if (state_id == AUTO_PILOT) {
+            return AutoPilot;
+        }
+        else if (state_id == MANUAL_OPERATION_WAITING) {
+            return ManualControl;
+        } 
+    }
+
+    return Idle;
+}
+
+State_t StateMachineThread::AutoPilotEvent(void)
```

Comment:
ya so for the autopilot algorithm, it essentially should just set the `target_speed` variable of the `global_mc_data` object to the max speed until some X distance before the `track_length` is hit.

All this data should be accessible from that global object. If not feel free to add them.

For the max speed, this is currently unknown, so just add a `#define` for it and we can go back and change it later.

Reply:
Thanks, I was just wondering do you want to me to mutate the object attribute like `global_mc_data.target_speed = MAX`. If so, I talked to Dev earlier and he said not to do this and it should be set in the PID thread which has not been implemented yet

Reply:
Ya what you described should be good. It doesn't really require a mutex since only this thread is going to be writing to that particular class member